### PR TITLE
Autocomplete suggestions for tags

### PR DESCRIPTION
**Pull request by solinger10**
This addresses #1999. I got this first running using a simple array adapter instead of the WordPress suggestions:

    String[] test = {"Test 1","Test 2","Test 3", "A Test","B Test","C Test"};
    ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, test);
    mTags.setAdapter(adapter);

After adding suggestions using SuggestionUtils, I wasn't sure about how the suggestions list works. Is there just one suggestions list per blog and it is the list that should be used for tag recommendations in this case? And are tags submitted ever added to this suggestion list? When running it on my test blog it seemed like the only thing in the suggestion list was my username.

### CODE REVIEW

**Comment by maxme** [comment]
`SuggestionAutoCompleteText` can be reused but you'll have to feed it with a different dataset. 

The adapter returned by `SuggestionUtils.setupSuggestions()` read data from a db table (`SuggestionTable`) that is filled by `SuggestionService`. Current implementation of `SuggestionService` only fetches username list for a certain blog.

In order to fix #1999, you need to make that `SuggestionService ` more generic to fetch tags from a certain WordPress site and store these tags in a db table (IMO, `SuggestionTable` should be reused too but you can create a separate table). Then you'll have to create the same kind of function `SuggestionUtils.setupSuggestions()` that bind the tags dataset to the adapter and use [here](https://github.com/solinger10/WordPress-Android/blob/6c89ff8dc355c2910d3d72466fbe1bdb38ea26a3/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L238).

**Comment by solinger10** [comment]
Thanks, makes sense.

**Comment by solinger10** [comment]
Okay, it now fetches and caches tag lists for sites. Right now it shows any tags on the current site that contain a word that starts with the text being typed.

Here is what it looks like:

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -220,6 +228,17 @@ public boolean onCreateOptionsMenu(Menu menu) {
         super.onCreateOptionsMenu(menu);
         MenuInflater inflater = getMenuInflater();
         inflater.inflate(R.menu.edit_post, menu);
+
+        mTags = (SuggestionAutoCompleteText) findViewById(R.id.tags);
+        mTags.setTokenizer(new SuggestionAutoCompleteText.CommaTokenizer());
+
+        remoteBlogId = WordPress.getCurrentRemoteBlogId();
+        mSuggestionServiceConnectionManager = new SuggestionServiceConnectionManager(this, remoteBlogId);
+        mTagSuggestionAdapter = SuggestionUtils.setupTagSuggestions(remoteBlogId, this, mSuggestionServiceConnectionManager);
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Bind service is called here but unbind is never called. Should be bound/unbound during onResume/onPause.

File: WordPress/src/main/java/org/wordpress/android/ui/suggestion/service/SuggestionService.java [hunk 2]
```diff
@@ -97,11 +100,68 @@ public void run() {
...
+            @Override
+            public void onErrorResponse(VolleyError volleyError) {
+                AppLog.e(AppLog.T.SUGGESTION, volleyError);
+                removeSiteIdFromTagRequestsAndStopServiceIfNecessary(remoteBlogId);
+            }
+        };
+
+        AppLog.d(AppLog.T.SUGGESTION, "suggestion service > updating tags for siteId: " + remoteBlogId);
+        String path = "/sites/" + remoteBlogId + "/tags";
+        WordPress.getRestClientUtils().get(path, listener, errorListener);
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/suggestion/service/SuggestionService.java]: That won't work on self hosted sites (they only support XMLRPC API calls). We should fill a new issue for this.

**Comment by solinger10** [comment]
What should be done to handle the self hosted sites?

**Comment by maxme** [comment]
> What should be done to handle the self hosted sites?

Communication between self hosted sites and the app is done with the WordPress XMLRPC API https://codex.wordpress.org/XML-RPC_WordPress_API To handle self hosted sites, we must have a fallback call to XMLRPC if the site is not wpcom-REST able (if the site is not a wpcom site or a jetpack powered self hosted).

**Comment by maxme** [comment]
@solinger10 we're going to drop Autocomplete tags support for self hosted and call this a wpcom/JetPack only feature - hope you haven't started to work on it.

File: WordPress/src/main/java/org/wordpress/android/ui/suggestion/service/SuggestionService.java [hunk 3]
```diff
@@ -97,11 +100,68 @@ public void run() {
...
+
+                JSONArray jsonTags = jsonObject.optJSONArray("tags");
+                List<Tag> tags = Tag.tagListFromJSON(jsonTags, remoteBlogId);
+                if (tags != null) {
+                    SuggestionTable.insertTagsForSite(remoteBlogId, tags);
+
+                    LocalBroadcastManager lbm = LocalBroadcastManager.getInstance(WordPress.getContext());
+                    Intent intent = new Intent(SuggestionService.ACTION_TAGS_LIST_UPDATED);
+                    intent.putExtra(SuggestionService.TAGS_LIST_UPDATED_EXTRA, remoteBlogId);
+                    lbm.sendBroadcast(intent);
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/suggestion/service/SuggestionService.java]: This event is broadcasted but never caught. We should catch it and update `mTagSuggestionAdapter` like it was done for name suggestion here for instance: https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L381-L385

**Comment by solinger10** [comment]
Sounds good, I had no communication for the past week as I was out of the country and had no internet. Seems as though I messed things up right when I got back to it. I was trying to get the branch up to date and accidentally merged in master instead of develop. Any chance you can help me out? I also tried reverting it as you can see, but still can't get it to work right now.

**Comment by maxme** [comment]
I merged your branch and `develop` in `/issue/1999-tag-list-autocomplete`.

Please rebase your branch to remove the `master` merge and revert.

**Comment by solinger10** [comment]
Ah I see that now. I removed the merge.

**Comment by maxme** [comment]
Merged via `issue/1999-tag-list-autocomplete` - this will ship with 4.1.

**Pull request merged**